workflow/controller/controller.go

```diff
@@ -535,19 +535,63 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
 				return err
 			}
 		case labelPodCompleted:
-			_, err := pods.Patch(
+			patch := []common.DeleteFinalizerPatch{
+				{
+					Operation: "replace",
+					Path:      "/metadata/labels/workflows.argoproj.io~1completed",
+					Value:     "true",
+				},
+			}
+			p, err := wfc.getPodFromAPI(ctx, namespace, podName)
+			if err != nil {
+				return err
+			}
+			i := slices.Index(p.Finalizers, common.Finalizer)
+			if i >= 0 {
+				patch = append(patch, common.DeleteFinalizerPatch{
+					Operation: "remove",
+					Path:      fmt.Sprintf("/metadata/finalizers/%d", i),
+				})
+			}
+			data, _ := json.Marshal(patch)
+			log.WithField("data", string(data)).Debug("patching pod")
+			_, err = pods.Patch(
 				ctx,
 				podName,
-				types.MergePatchType,
-				[]byte(`{"metadata": {"labels": {"workflows.argoproj.io/completed": "true"}}}`),
+				types.JSONPatchType,
+				data,
 				metav1.PatchOptions{},
 			)
 			if err != nil {
 				return err
 			}
 		case deletePod:
+			var patch []common.DeleteFinalizerPatch
+			p, err := wfc.getPodFromAPI(ctx, namespace, podName)
+			if err != nil {
+				return err
+			}
+			i := slices.Index(p.Finalizers, common.Finalizer)
+			if i >= 0 {
+				patch = append(patch, common.DeleteFinalizerPatch{
+					Operation: "remove",
+					Path:      fmt.Sprintf("/metadata/finalizers/%d", i),
+				})
+			}
+			data, _ := json.Marshal(patch)
+			log.WithField("data", string(data)).Debug("patching pod")
+			_, err = pods.Patch(
+				ctx,
+				podName,
+				types.JSONPatchType,
+				data,
+				metav1.PatchOptions{},
+			)
+			if err != nil {
+				return err
+			}
 			propagation := metav1.DeletePropagationBackground
-			err := pods.Delete(ctx, podName, metav1.DeleteOptions{
+			err = pods.Delete(ctx, podName, metav1.DeleteOptions{
```

Comment:
I assume that in general, a given Pod reaches `labelPodCompleted` and then `deletePod` logic? Do you know if it's possible for that not to happen?

Reply:
The execution of the deletePod logic is decided by the following function.
https://github.com/argoproj/argo-workflows/blob/5b6ad2be163ecd3f0251a931ab84dba3c6085ad2/workflow/controller/pod_cleanup.go#L46-L72

This belongs by Pod GC strategy.
https://argoproj.github.io/argo-workflows/fields/#:~:text=pod%20GC%20queue.-,strategy,-string
https://github.com/argoproj/argo-workflows/blob/dbf4215e6f25765517ca6534e9eaee13995cdd1b/examples/pod-gc-strategy.yaml#L10-L15